File: as1/q1.js
/*
Q1. Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.
You may assume that each input would have exactly one solution, and you may not use the same element twice.
You can return the answer in any order.

Input: nums = [2,7,11,15], target = 9
Output: [0,1]
Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
*/

var twoSum = function(nums, target) {
    const hashmap = {};
    for (let i = 0; i < nums.length; i++) {
      const diff = target - nums[i];
  
      // If the difference is in the hashmap, return the indices.
      if (diff in hashmap) { // or, hashmap[diff] !== undefined
        return [hashmap[diff], i];
      }
  
      // Otherwise, add the current number to the hashmap.
      hashmap[nums[i]] = i;
    }
  
    return [];
  };

let nums = [1,2,4,3,5];
let target = 8;

console.log(twoSum(nums, target))

File: as3/q1.cpp
/*
Question 1
Given an integer array nums of length n and an integer target, find three
integers in nums such that the sum is closest to the target. Return the sum of
the three integers.

You may assume that each input would have exactly one solution.

Example 1:
Input: nums = [-1,2,1,-4], target = 1
Output: 2

Explanation: The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).
*/



#include <bits/stdc++.h>

using namespace std;

int threeSumClosest(vector<int> nums, int target) {
    sort(nums.begin(), nums.end());

    int closest = 0;
    int minAbs = INT_MAX;
    for (int i = 0; i < nums.size() - 2; i++) {
        int low = i + 1;
        int high = nums.size() - 1;

        while (low < high) {
            int sum = nums[i] + nums[low] + nums[high];

            if (abs(target - sum) < minAbs) {
                minAbs = abs(target - sum);
                closest = sum;
            }

            if (sum == target) {
                return target;
            } else if (sum > target) {
                high= high - 1;
            } else {
                low = low + 1;
            }
        }
    }

    return closest;
}

int main() { 
    cout<<threeSumClosest({-1,2,1,-4},1);
    return 0; 
}